sos/plugins/juju.py

```diff
@@ -9,113 +9,201 @@
+        # Make sure it looks like juju is configured before continuing
+        username = self.get_option("juju-user")
+        homedir = os.path.expanduser("~" + username)
+        if os.path.exists(homedir + "/.local/share/juju"):
+            self.collect_juju_output(username)
+
+    def postproc(self):
+        juju_config_keys = [
+            "password",
+            "credentials",
+            "client_password",
+            "endpoint-tls-ca",
+            "docker-logins",
+            "license-key",
+            "registry-credentials",
+        ]
+
+        juju_config_certs = ["ca-cert"]
+
+        # Handle the case where the key and value are on seperate lines
+        # in a juju config yaml file
+        juju_config_regex = (
+            r"(?m)^(\s*)(((%s):((\n\1\s+.*)|\n)+)\1\s+value:\s*).*"
+            % "|".join(juju_config_keys)
+        )
+        juju_config_sub_regex = r"\1\2*********"
+
+        # Will match and replace a certificat in a yaml file
+        certs_regex = (
+            r"((?m)^\s*(%s)\s*:\s*)(\|\s*\n\s+-+BEGIN (.*)"
+            r"-+\s(\s+\S+\n)+\s+-+END )\4(-+)" % "|".join(juju_config_certs)
+        )
+        certs_sub_regex = r"\1*********"
+
+        # Go through all output files and sanitize them (sos doesn't have a
+        # builtin method of doing this currently):
+        output_dir = self.get_cmd_output_path(make=False)
+        juju_config_files = glob.glob(output_dir + "/juju_config*.yaml")
+        juju_cert_files = glob.glob(output_dir + "/juju*controller*.yaml")
+
+        for file in juju_config_files:
+            self._log_debug(
+                "substituting {} for {} in {}".format(
+                    juju_config_sub_regex, juju_config_regex, file
+                )
+            )
+            try:
+                fp = open(file, "r")
+                result, replacements = re.subn(
+                    juju_config_regex, juju_config_sub_regex, fp.read()
+                )
+                fp.close()
+                fp = open(file, "w")
+                fp.write(result)
+                fp.close()
+            except Exception as e:
+                msg = "regex substitution failed for '%s' with: '%s'"
+                self.log_error(msg % (file, e))
+
+        for file in juju_cert_files:
+            self._log_debug(
+                "substituting {} for {} in {}".format(
+                    certs_sub_regex, certs_regex, file
+                )
+            )
+            try:
+                fp = open(file, "r")
+                result, replacements = re.subn(
+                    certs_regex, certs_sub_regex, fp.read()
+                )
+                fp.close()
+                fp = open(file, "w")
+                fp.write(result)
+                fp.close()
+            except Exception as e:
+                msg = "regex substitution failed for '%s' with: '%s'"
+                self.log_error(msg % (file, e))
```

Comment:
Are you kidding me? If you don't find what you need in the canned output substitution methods provided, talk to us - we don't just throw dozens of lines of ad-hoc open code into plugins like this. It is neither sustainable nor maintainable.